src/hotspot/share/c1/c1_LIR.hpp

```diff
@@ -2034,7 +2034,7 @@ class LIR_OpProfileCall : public LIR_Op {
   bool should_profile_receiver_type() const {
     bool callee_is_static = _profiled_callee->is_loaded() && _profiled_callee->is_static();
     Bytecodes::Code bc = _profiled_method->java_code_at_bci(_profiled_bci);
-    bool call_is_virtual = (bc == Bytecodes::_invokevirtual && !_profiled_callee->can_be_statically_bound()) || bc == Bytecodes::_invokeinterface;
+    bool call_is_virtual = (bc == Bytecodes::_invokevirtual && (UseCHA || !_profiled_callee->can_be_statically_bound())) || bc == Bytecodes::_invokeinterface;
```

Comment:
UseCHA is on by default, so this effectively always turns on profiling.  I think it would be better if we set a flag that says if CHA was used or not.

Reply:
IMO a better option is to simply drop `can_be_statically_bound()` check. As this bug demonstrates, selected method at call site (represented by `_profiled_callee`) is not an accurate representation of call site behavior.

Instead, to keep things simple, a check for private methods (which don't participate in virtual dispatch) can be introduced.

Reply:
Would it be more appropriate to use can_be_statically_bound(ciInstanceKlass* context) here, where in this case _profiled_callee is Child1 and context is Parent?

Reply:
That's definitely an option, but then you need to determine the context. Do you propose to use declared method holder as a context here?

Reply:
Yes, exactly.

Reply:
I am not sure how Dean proposal will help. I agree with Vladimir's suggestion - C1 should not optimize call sites in Level 3 compilation.

Reply:
I was just trying to think how to preserve the original intend of this code, which seemed to be "skip profiling if the profiling info is not needed", but getting it right seems complicated, so I'm OK with always doing it.

> C1 should not optimize call sites in Level 3 compilation.

You mean don't use can_be_statically_bound() and other checks to devirtualize virtual calls?  I believe C1 does that at all compilation levels.

Reply:
It does look attractive to align the logic with C2 usage of type profiles (avoid profiling when C2 doesn't consume the data). But I feel more comfortable unifying different modes of profiling at the expense of some micro-optimization opportunities. In other words, if interpreter collects some bit of data, I'd prefer to see C1 doing the same (and vice versa). 

I took a look at interpreter code (in `TemplateTable::invokevirtual_helper()`) and it makes the decision at runtime based on `is_vfinal` flag on `ResolvedMethodEntry`. The flag is set in `ConstantPoolCache::set_direct_or_vtable_call()` and covers both private and final methods. Moreover, receiver profiling is not performed on `invokeinterface` of private methods which is not taken into account by `should_profile_receiver_type()` now.

It looks tempting to replicated what interpreter does (inspect `vfinal` flag on resolved method), but C1 has to gracefully work with not-yet-resolved call sites. So, either a recompilation or a runtime check is needed to align the behavior with interpreter.

I haven't looked into the details, but performing profiling in C1 when the rest of the JVM doesn't expect that makes me a bit nervous. Smells like a possible source of profile data corruption.

Reply:
What profiling can be done seems to be decided by MethodData::compute_data_size()/MethodData::initialize_data(), which uses profile_arguments_for_invoke() and profile_return_for_invoke().  At runtime, I believe profiling is restricted by what it finds in the MethodData.

Reply:
Thanks for additional details, Dean. Thinking more about it, it looks like the worst case scenario possible is that there'll be more data gathered than needed, but there'll be always slots reserved for the data.
IMO current fix seems like a good compromise.